feat: Add seven-segment counter system

nadasuhailAyesh12 opened the pull request:
- Implemented a system to display a decrementing 3-digit number (555 to 000) on a seven-segment display using an LPC11xx microcontroller.
  - Configured GPIO ports for input (digit selectors) and output (seven-segment display and red LED).
  - Added functionality to control the display of hundreds, tens, and ones places.
  - Red LED turns on when the counter reaches 0.
  - Ensured proper initialization of LEDs and GPIO ports.

- Added a detailed PDF report documenting:
  - Problem description and solution overview.
  - Code structure and design rationale.
  - Component connections and simulation results with observations.

This commit includes the full implementation and report for Lab 2 of the Embedded Systems course.

nadasuhailAyesh12 committed:
feat: Add seven-segment counter system 


- Implemented a system to display a decrementing 3-digit number (555 to 000) on a seven-segment display using an LPC11xx microcontroller.
  - Configured GPIO ports for input (digit selectors) and output (seven-segment display and red LED).
  - Added functionality to control the display of hundreds, tens, and ones places.
  - Red LED turns on when the counter reaches 0.
  - Ensured proper initialization of LEDs and GPIO ports.

- Added a detailed PDF report documenting:
  - Problem description and solution overview.
  - Code structure and design rationale.
  - Component connections and simulation results with observations.

This commit includes the full implementation and report for Lab 2 of the Embedded Systems course.

adabowsa commented on lab2/main.c:
```diff
@@ -0,0 +1,61 @@
+#include <LPC11xx.h>
+
+// Define GPIO registers
+#define GPIO0DIR (*((volatile unsigned long *)0x50008000))
```
Use uint32_t for Register Definitions:

adabowsa commented on lab2/lab2Report.pdf:
Adding comments to explain the purpose of each section of the code can improve readability.

adabowsa commented on lab2/main.c:
```diff
@@ -0,0 +1,61 @@
+        // Handle hundreds place
+        GPIO0DATA = 0x04;  // Select the hundreds digit
+        GPIO2DATA = seven_seg_encoder[num / 100]; // Display hundreds place
+        for (i = 0; i < 5000; i++); // Delay
```
Consider using a hardware timer to create precise delays instead of using busy-wait loops.

nadasuhailAyesh12 committed:
Refactor GPIO register definitions to use uint32_t

use uint32_t instead of long  for clarity and consistency

nadasuhailAyesh12 committed:
Refactor delay mechanism: Replace busy-wait loops with hardware timer (Timer32B0)

- Configured Timer32B0 to handle delays using match interrupts.
- Implemented `Timer32B0_Start` and `Timer32B0_Wait` functions for precise timing.
- Replaced busy-wait loops with timer-based delays for better efficiency and accuracy.
- Ensured delays remain consistent (~5 ms) by calculating match values based on a 48 MHz clock.
- Reduced CPU utilization during delays, allowing for potential multitasking or low-power modes

nadasuhailAyesh12 committed:
Adding more commits to improve readability
